### DISCUSSION TITLE
Can't get this lib to cascade-save a secondary relationship

### DISCUSSION THREAD

**Discussion by haf** [sqlalchemy/sqlalchemy#8007] [Usage Questions] [answered]
Aim; to have full cascade save/load/delete with SQLAlchemy from the root entity (scenario) through Segment and then Timeseries.

Question; from a parent class, BaseSegment, with multiple properties that link to a child class, Timeseries, whose table links back with segment_id, how do I configure a backref?

```python
class Timeseries(Base):
    __tablename__ = "timeseries"
    __table_args__ = {"schema": "public"}
    # values etc
    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=FetchedValue())
    segment_id = Column(UUID(as_uuid=True), ForeignKey("public.segments.id"))

class BaseSegment(Base):
    __tablename__ = "segments"
    __table_args__ = {"schema": "public"}
    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=FetchedValue())
    actual_effect_id = Column(UUID(as_uuid=True), ForeignKey(Timeseries.id))
    actual_effect = relationship(
        Timeseries,
        # back_populates="segment",
        backref="segment",
        # foreign_keys=[actual_effect_id],
        primaryjoin=foreign(actual_effect_id) == remote(Timeseries.segment_id),
        # remote_side=Timeseries.segment_id,
        uselist=False,
    )
    scenario_id = Column(UUID(as_uuid=True), ForeignKey("public.scenarios.id"))
    scenario = relationship("Scenario")

class Scenario(Base):
    __tablename__ = "scenarios"
    __table_args__ = {"schema": "public"} # https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#explicit-schema-name-with-declarative-table
    id = Column(UUID(as_uuid=True), primary_key=True, index=True, server_default=FetchedValue())
    segments = relationship("BaseSegment", back_populates="scenario") # this one seems to work


#  elsewhere

async def create(db: Session, sc: Scenario, created_by: UUID) -> crud_models.Scenario:
    scenario = crud_models.Scenario.of_model(sc, created_by)
    db.add(scenario)
    db.commit()
    db.refresh(scenario)
    return scenario

```

Crashes like this:

> sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class BaseSegment->segments'. Original exception was: **Relationship BaseSegment.actual_effect could not determine any unambiguous local/remote column pairs based on join condition and remote_side arguments**.  Consider using the remote() annotation to accurately mark those elements of the join condition that are on the remote side of the relationship.

In particular, when I have `remote_side=Timeseries.segment_id` I get this crash.

If I remove a bit of the config above, instead of the Session cascade-writing from segment -> timeseries, I get this error\*:

> sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "segment_id" of relation "timeseries" violates not-null constraint

Furthermore, if I generate the UUID id from the calling python code, the DB complains that basically the segment's Id doesn't exist (yet!)\**.

I can't just do a `back_populates` because I don't know whether the Timeseries is for a `predicted_effect` or `actual_effect`. How do I model this?

When using `back_populates` I can easily trigger this exception:

> sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class BaseSegment->scenario_segments'. Original exception was: Timeseries.segment and back-reference BaseSegment.actual_effect are both of the same direction symbol('MANYTOONE').  Did you mean to set remote_side on the many-to-one side ?

But this is not a self-referential relationship.

---

\***Details, full error:**

```
E       sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "segment_id" of relation "timeseries" violates not-null constraint
E       DETAIL:  Failing row contains (b2dff54f-ff1c-4457-8cf1-0783ea2acdbb, effect_var, null, revenue, EUR, Resulting amount revenue per day, {45753.00000000003,63875.00000000002,12731.999999999998,56585,37..., datetime64[ns], D, null, null, null, revenue, null).
E
E       [SQL: INSERT INTO public.timeseries (id, type, variable_name, variable_unit, variable_description, values, dtype, freq, segment_id, funnel_level, treatment, behavioural_segment_id, effect) VALUES (%(id)s, %(type)s, %(variable_name)s, %(variable_unit)s, %(variable_description)s, %(values)s::FLOAT[], %(dtype)s, %(freq)s, %(segment_id)s, %(funnel_level)s, %(treatment)s, %(behavioural_segment_id)s, %(effect)s)]
E       [parameters: {'id': UUID('b2dff54f-ff1c-4457-8cf1-0783ea2acdbb'), 'type': 'effect_var', 'variable_name': 'revenue', 'variable_unit': 'EUR', 'variable_description': 'Resulting amount revenue per day', 'values': [45753.00000000003, 63875.00000000002, 12731.999999999998, 56585.0, 37048.99999999997, 19938.999999999985, 28346.999999999985], 'dtype': 'datetime64[ns]', 'freq': 'D', 'segment_id': None, 'funnel_level': None, 'treatment': None, 'behavioural_segment_id': None, 'effect': 'revenue'}]
```

\**For the client-generated UUID:

```
sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) insert or update on table "timeseries" violates foreign key constraint "timeseries_segment_id_fkey"
```

Segments FK:s:
<img width="507" alt="image" src="https://user-images.githubusercontent.com/193115/167820427-2b81ecb9-35e2-4bda-8382-99542cc4745c.png">

Timeseries FK:s
<img width="532" alt="image" src="https://user-images.githubusercontent.com/193115/167820504-3b9e593a-5502-4f0a-92fb-148f85de4eff.png">

### References

- [backref](https://docs.sqlalchemy.org/en/14/orm/backref.html)
- [working with related objects](https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html)
- [annotations recommended](https://docs.sqlalchemy.org/en/14/glossary.html#term-annotations)
- [mailing list similar error](https://groups.google.com/g/sqlalchemy/c/Ir7P5CkOijU)

**Comment by zzzeek** [comment]
Im adding another fix to my example, will repost in a moment.  you have mutualyl dependent foreign keys so this also needs [post_update](https://docs.sqlalchemy.org/en/14/orm/relationship_persistence.html?highlight=mutually#rows-that-point-to-themselves-mutually-dependent-rows)

**Comment by zzzeek** [comment, marked as answer]
OK so the intent of this model is not totally clear as you have mutually referring foreign key constraints between BaseSegment and Timeseries, which means that a list of Timeseries objects can all refer to one BaseSegment, but at the same time BaseSegment can additionally point to a "single" Timeseries object.   This architecture is described at https://docs.sqlalchemy.org/en/14/orm/relationship_persistence.html?highlight=mutually#rows-that-point-to-themselves-mutually-dependent-rows where we have the concept of a "list of things" and a "favorite thing" inside that list.     So if that's what you want, you need to use the post_update flag on the persistence side, and additionally you will want names on the foreign keys as I have below so that the test script can run "drop" / "create" statements and know which foreign keys need to be dropped separately.  remote_side / primaryjoin are not needed, foreign_keys is enough to resolve the ambiguity (though the settings you had worked fine on this end also).

additionally, you can't use UUID and then server_default=FetchedValue because there's no server-side Postgresql function specified.   The schema needs a real default generator if you don't want to generate the UUID values in Python; [get_random_uuid()](https://www.postgresql.org/docs/current/functions-uuid.html) can be used for this.     Finally you dont want to create indexes on single-col primary keys because there is already an index on a primary key.


Working example that populates all the columns is below.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class Timeseries(Base):
    __tablename__ = "timeseries"
    # values etc
    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    segment_id = Column(UUID(as_uuid=True), ForeignKey("segments.id"))


class BaseSegment(Base):
    __tablename__ = "segments"
    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    actual_effect_id = Column(
        UUID(as_uuid=True), ForeignKey(Timeseries.id, name="fk_tid")
    )
    actual_effect = relationship(
        Timeseries,
        backref="segment",
        foreign_keys=[actual_effect_id],
        post_update=True,
    )
    scenario_id = Column(UUID(as_uuid=True), ForeignKey("scenarios.id"))
    scenario = relationship("Scenario", back_populates="segments")

    timeseries = relationship(Timeseries, foreign_keys=[Timeseries.segment_id])


class Scenario(Base):
    __tablename__ = "scenarios"
    id = Column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=func.gen_random_uuid(),
    )
    segments = relationship(
        "BaseSegment", back_populates="scenario"
    )


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

with Session(e) as sess:
    ts1 = Timeseries()
    ts2 = Timeseries()
    base_segment = BaseSegment(timeseries=[ts1, ts2], actual_effect=ts1)

    scenario = Scenario(segments=[base_segment])
    sess.add(scenario)
    sess.commit()
    sess.refresh(scenario)
```

output, note the UPDATE at the end to set "actual_effect_id", as well as that ALTER is used to add the foreign key constraints since there is a reference cycle between them:

```
2022-05-11 08:44:12,536 INFO sqlalchemy.engine.Engine 
CREATE TABLE timeseries (
	id UUID DEFAULT gen_random_uuid() NOT NULL, 
	segment_id UUID, 
	PRIMARY KEY (id)
)


2022-05-11 08:44:12,536 INFO sqlalchemy.engine.Engine [no key 0.00007s] {}
2022-05-11 08:44:12,541 INFO sqlalchemy.engine.Engine 
CREATE TABLE segments (
	id UUID DEFAULT gen_random_uuid() NOT NULL, 
	actual_effect_id UUID, 
	scenario_id UUID, 
	PRIMARY KEY (id)
)


2022-05-11 08:44:12,542 INFO sqlalchemy.engine.Engine [no key 0.00039s] {}
2022-05-11 08:44:12,549 INFO sqlalchemy.engine.Engine 
CREATE TABLE scenarios (
	id UUID DEFAULT gen_random_uuid() NOT NULL, 
	PRIMARY KEY (id)
)


2022-05-11 08:44:12,549 INFO sqlalchemy.engine.Engine [no key 0.00029s] {}
2022-05-11 08:44:12,554 INFO sqlalchemy.engine.Engine ALTER TABLE segments ADD FOREIGN KEY(scenario_id) REFERENCES scenarios (id)
2022-05-11 08:44:12,554 INFO sqlalchemy.engine.Engine [no key 0.00042s] {}
2022-05-11 08:44:12,557 INFO sqlalchemy.engine.Engine ALTER TABLE segments ADD CONSTRAINT fk_tid FOREIGN KEY(actual_effect_id) REFERENCES timeseries (id)
2022-05-11 08:44:12,557 INFO sqlalchemy.engine.Engine [no key 0.00019s] {}
2022-05-11 08:44:12,559 INFO sqlalchemy.engine.Engine ALTER TABLE timeseries ADD FOREIGN KEY(segment_id) REFERENCES segments (id)
2022-05-11 08:44:12,559 INFO sqlalchemy.engine.Engine [no key 0.00017s] {}
2022-05-11 08:44:12,560 INFO sqlalchemy.engine.Engine COMMIT
2022-05-11 08:44:12,571 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-05-11 08:44:12,572 INFO sqlalchemy.engine.Engine INSERT INTO scenarios DEFAULT VALUES RETURNING scenarios.id
2022-05-11 08:44:12,572 INFO sqlalchemy.engine.Engine [generated in 0.00014s] {}
2022-05-11 08:44:12,574 INFO sqlalchemy.engine.Engine INSERT INTO segments (actual_effect_id, scenario_id) VALUES (%(actual_effect_id)s, %(scenario_id)s) RETURNING segments.id
2022-05-11 08:44:12,574 INFO sqlalchemy.engine.Engine [generated in 0.00013s] {'actual_effect_id': None, 'scenario_id': UUID('86bf8054-3de1-4d2b-a26a-f2f5350cb72f')}
2022-05-11 08:44:12,575 INFO sqlalchemy.engine.Engine INSERT INTO timeseries (segment_id) VALUES (%(segment_id)s) RETURNING timeseries.id
2022-05-11 08:44:12,575 INFO sqlalchemy.engine.Engine [generated in 0.00011s] ({'segment_id': UUID('5272556f-5490-4cd5-9b3f-94aab284b7cc')}, {'segment_id': UUID('5272556f-5490-4cd5-9b3f-94aab284b7cc')})
2022-05-11 08:44:12,577 INFO sqlalchemy.engine.Engine UPDATE segments SET actual_effect_id=%(actual_effect_id)s WHERE segments.id = %(segments_id)s
2022-05-11 08:44:12,577 INFO sqlalchemy.engine.Engine [generated in 0.00011s] {'actual_effect_id': UUID('70b283ed-5275-4d12-bef5-a101f2f559a0'), 'segments_id': UUID('5272556f-5490-4cd5-9b3f-94aab284b7cc')}
2022-05-11 08:44:12,578 INFO sqlalchemy.engine.Engine COMMIT
2022-05-11 08:44:12,580 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-05-11 08:44:12,581 INFO sqlalchemy.engine.Engine SELECT scenarios.id 
FROM scenarios 
WHERE scenarios.id = %(pk_1)s
2022-05-11 08:44:12,581 INFO sqlalchemy.engine.Engine [generated in 0.00011s] {'pk_1': UUID('86bf8054-3de1-4d2b-a26a-f2f5350cb72f')}
2022-05-11 08:44:12,582 INFO sqlalchemy.engine.Engine ROLLBACK

```

**Reply by haf** [reply]
Thank you, the trick was to additionally have a `timeseries` collection and also `post_update=True`!

**Reply by haf** [reply]
So, bumping into a follow up here; with `post_update` there's no pre-delete that clears this field?

Found another identical issue here https://stackoverflow.com/questions/50356012/sqlalchemy-pre-delete-update-not-running-when-post-update-true

Getting

```
E           sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "scenario_timeseries" violates foreign key constraint "scenario_segments_predicted_from_current_id_fkey" on table "scenario_segments"
E           DETAIL:  Key (id)=(6d874d97-e2fd-4e3a-a521-d5c15a3d7908) is still referenced from table "scenario_segments".
E
E           [SQL: DELETE FROM public.scenario_timeseries WHERE public.scenario_timeseries.id = %(id)s]
E           [parameters: ({'id': UUID('6d874d97-e2fd-4e3a-a521-d5c15a3d7908')}, {'id': UUID('6f794ac0-d1ec-4329-89a8-5716a150c6ba')})]
E           (Background on this error at: https://sqlalche.me/e/14/gkpj)
```

**Reply by zzzeek** [reply]
post_update implies that DELETE should also do the right thing.   

taking my example above and adding:

```py
with Session(e) as sess:
    ts1 = Timeseries()
    ts2 = Timeseries()
    base_segment = BaseSegment(timeseries=[ts1, ts2], actual_effect=ts1)

    scenario = Scenario(segments=[base_segment])
    sess.add(scenario)
    sess.commit()
    sess.refresh(scenario)


    sess.delete(base_segment)
    sess.commit()
```


the SQL at the end succeeds and is below, where you can see two UPDATE statements setting None on the FK col.  the second UPDATE is not actually necessary but this is a limitation of post_update right now which is that it is overly eager about nulling out its FK column:

```sql
UPDATE timeseries SET segment_id=%(segment_id)s WHERE timeseries.id = %(timeseries_id)s
[{'segment_id': None, 'timeseries_id': UUID('159d4e6c-9552-48b5-bb0c-dbcc932d004b')}, {'segment_id': None, 'timeseries_id': UUID('d6e297ca-aefb-40d0-8965-b6c5bcdfee4e')}]

UPDATE segments SET actual_effect_id=%(actual_effect_id)s WHERE segments.id = %(segments_id)s
{'actual_effect_id': None, 'segments_id': UUID('fa0fcbf1-9af4-4abe-8acb-fbadc661c3a1')}

DELETE FROM segments WHERE segments.id = %(id)s
```

**Reply by haf** [reply]
My code is slightly different and that might account for the differences:

1. I don't want to have to track the collection of timeseries manually because I would like the mapper to do that (and it's not easy), so I try to make changes cascade down from the scenario:
   ```python
    async def update(db: Session, sc: models.Scenario) -> models.Scenario:
        logger.info("Updating a Scenario entity")

        scenario = crud_models.Scenario.of_model(sc)
         _pretty_print_scenario('update', scenario, logger.debug)

        logger.debug("Merging updates into UoW / db", {
        "segments": list(map(lambda x: x.id, scenario.segments))
        })
        db.merge(scenario)
        db.flush()

        return scenario.to_model()
    ```
 2. The caller looks like this:
    ```python
    scenario = crud.read() -> models.Scenario (not a crud_models.*)
    change lots of stuff on it, primarily in the timeseries in a way that is known to the "domain logic"
    domain.compute(scenario) -> mutates it
    crud.update(scenario)
    ```

In the end I've removed two timeseries from the segment as they were created as new instances; one effect timeseries and one treatment timeseries. These get new Ids, and are added to the timeseries-collection as well as set as the properties (like is shown in this screenshot, highlighted a new instance without a DB-generated Id assigned yet)

<img width="927" alt="image" src="https://user-images.githubusercontent.com/193115/170011509-5d3fc970-e8b5-4d00-8420-4863337a6edb.png">

and also at the bottom as the timeseries with a None id (to be persisted)

For this sqlalchemy generates:

1. 1 read from scenarios
2. 1 read from segments
3. 1 read from annotations (another relation)
4. 1 read from timeseries
5. 2 reads from annotations outer joined with all three items that can be annotated: scenario, segment and timeseries, the second only for the timeseries that was deleted

It then procedes to updating:

1. update scenarios setting a single property timezone
2. update segments settings two properties, start and end
3. update timeseries that was not changed (as far as I can tell) (metadata properties only)
4. update timeseries that was changed with new values 😀 (treatment ts)
5. update timeseries that was not changed (as far as I can tell) (metadata properties only)
6. update 9 timeseries that was not changed (as far as I can tell) (metadata properties only)
7. update timeseries that was not changed (as far as I can tell) (metadata properties only)
8. insert the new timeseries instance (effect ts) 😀

Here, however, the old one is deleted, and no post-update is issued to set the property on Segment (as this was a replacement operation), and it does not pre-update the FK from segment before deleting.

<img width="1673" alt="image" src="https://user-images.githubusercontent.com/193115/170020869-2b2a4ed1-c356-42a6-b9cd-5cadcc448123.png">

So when read back, the instance is None from Segment->Timeseries:

<img width="1075" alt="image" src="https://user-images.githubusercontent.com/193115/170021067-ef702067-421d-4a57-814e-aab56015bd89.png">

Would it therefore be right to assume that merge doesn't properly do the post-update / pre-delete hooks?

**Reply by zzzeek** [reply]
hey there -

that's lots of details without any code to attach it towards.   Can you simply take my example above and modify it to show how you get your error?  then ill know what pattern you are trying to do.
